emulator/src/extensions/gamepad/ext_gamepad.c

```diff
@@ -0,0 +1,628 @@
+
+                        event->type   = BUTTON;
+                        event->button = buttonIdx;
+
+                        if (new->dwButtons & (1 << buttonIdx))
+                        {
+                            // Button pressed
+                            event->value = 1;
+                            cur->dwButtons |= (1 << buttonIdx);
+                        }
+                        else
+                        {
+                            // Button released
+                            event->value = 0;
+                            cur->dwButtons &= ~(1 << buttonIdx);
+                        }
+
+                        applyEvent(joystick, event);
+                        return true;
+                    }
+                }
+                else if ((new->dwFlags& JOY_RETURNPOV) && cur->dwPOV != new->dwPOV)
+                {
+                    // This is kinda weird!
+                    // This is because we're splitting a single POV hat event into two discrete axes
+                    // So, we just send the X axis first, and then set a flag (and don't update the 'cur' value)
+                    // And next time if the flag is set, we return the Y axis instead, and then update the 'cur' value
+                    event->type = AXIS;
+                    if (!winData->pendingPov)
+                    {
+                        // Send this as the first axis (6)
+                        event->axis = 6;
+
+                        if (new->dwPOV == 65535)
+                        {
+                            event->value = 0;
+                        }
+                        else
+                        {
+                            event->value = CLAMP(-32 * getCos1024(((new->dwPOV / 100) + 90) % 360), -32768, 32767);
+                        }
+
+                        winData->pendingPov = true;
+                    }
+                    else
+                    {
+                        // Send this as the second axis (7)
+                        event->axis = 7;
+
+                        if (new->dwPOV == 65535)
+                        {
+                            event->value = 0;
+                        }
+                        else
+                        {
+                            event->value = CLAMP(-32 * getSin1024(((new->dwPOV / 100) + 90) % 360), -32768, 32767);
+                        }
+
+                        winData->pendingPov = false;
+                        cur->dwPOV          = new->dwPOV;
+                    }
+
+                    applyEvent(joystick, event);
+                    return true;
+                }
+                else
+                {
+                    // nothing is different!
+                    // check one more time
+                    winData->pendingState = false;
+                    continue;
+                }
+            }
+        } while (0);
+#elif defined(EMU_LINUX)
```

Comment:
Seems nicer

Reply:
I think so too but in fairness to windows I wrapped the linux API pretty tightly and just made the windows one comply with that which looks a bit more ugly lol